### PR TITLE
Convert public API data classes to Poko

### LINKED ISSUE #1043: Eliminate `data class` from public API
They're a nightmare to evolve in a source- and binary-compatible way.

Detekt has a rule for this!

### PR DESCRIPTION
**Pull request by drewhamilton**
Closes #1043. Depends on #1369.

Figured I'd get this up with the snapshot to confirm it's all working as expected before cutting Poko 0.15. Unfortunately it's blocked by Kotlin 1.9, unless we restore the reflection removed in drewhamilton/poko#196.

Took my best guess on which modules are public API, lmk if I missed anything!

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -15,13 +15,15 @@ buildscript {
...
     classpath libs.paparazzi.gradlePlugin
+    classpath libs.poko.gradlePlugin
     classpath 'app.cash.redwood.build:gradle-plugin'
     classpath 'app.cash.redwood:redwood-gradle-plugin'
   }
   repositories {
     mavenCentral()
     google()
     gradlePluginPortal()
+    maven { url = uri("https://oss.sonatype.org/content/repositories/snapshots") }
```
Comment (drewhamilton) [inline build.gradle:26]: TODO: Remove

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -74,5 +74,6 @@ coil-core = { module = "io.coil-kt:coil", version.ref = "coil" }
 turbine = "app.cash.turbine:turbine:1.0.0"
 ktlint = "com.pinterest:ktlint:0.50.0"
 googleJavaFormat = "com.google.googlejavaformat:google-java-format:1.17.0"
+poko-gradlePlugin = "dev.drewhamilton.poko:poko-gradle-plugin:0.15.0-SNAPSHOT"
```
Comment (drewhamilton) [inline gradle/libs.versions.toml:77]: TODO: Bump to release version

File: redwood-protocol/src/commonMain/kotlin/app/cash/redwood/protocol/protocol.kt [hunk 3]
```diff
@@ -48,33 +52,33 @@ public sealed interface Change {
...
+  public val value: JsonElement = JsonNull,
 ) : ValueChange
 
 @Serializable
 @SerialName("modifier")
-public data class ModifierChange(
+@Poko public class ModifierChange(
   override val id: Id,
-  val elements: List<ModifierElement> = emptyList(),
+  public val elements: List<ModifierElement> = emptyList(),
```
Comment (drewhamilton) [inline redwood-protocol/src/commonMain/kotlin/app/cash/redwood/protocol/protocol.kt]: TODO: Array

File: settings.gradle [hunk 4]
```diff
@@ -17,6 +17,7 @@ dependencyResolutionManagement {
   repositories {
     mavenCentral()
     google()
+    maven { url = uri("https://oss.sonatype.org/content/repositories/snapshots") }
```
Comment (drewhamilton) [inline settings.gradle:20]: TODO: Remove

**Comment by drewhamilton** [comment]
Most of CI fails with the reflection issue mentioned above. But `build` fails with this:

> Task :redwood-runtime:compileCommonMainKotlinMetadata FAILED
e: java.lang.IllegalStateException: e: Could not find "/.../redwood/redwood-runtime/build/kotlinTransformedMetadataLibraries/commonMain/dev.drewhamilton.poko-poko-annotations-0.15.0-SNAPSHOT-commonMain-oWB5zQ.klib" in [/.../Library/Application Support/kotlin/daemon]

Not sure if this is a configuration problem in this PR or something Poko is setting up wrong.

**Comment by JakeWharton** [comment]
If you rebase on #1369 the build gets much farther.

The problem seems to be that Poko's Gradle plugin has a `compileOnly` dependency on KGP 1.9 so it doesn't cause Gradle to upgrade here. The runtime, however, is built with Kotlin 1.9 metadata which fails to be resolved correctly with 1.8.22 for some reason.

**Comment by drewhamilton** [comment]
Ah OK, I'll rebase on the 1.9.0 PR. My thought behind using `compileOnly` was that a project should explicitly opt-in to upgraded Kotlin compiler versions, rather than be silently upgraded by Poko.

**Comment by JakeWharton** [comment]
Yeah it's a reasonable choice. Both options have upsides and downsides.

**Comment by drewhamilton** [comment]
Apparently KotlinX serialization omits empty lists when encoding, but includes empty arrays. Some of the tests fail as a result, and I'm not sure if they're safe to change (adding the empty array) or if more stuff needs to change along with them.

Examples below:
```
app.cash.redwood.protocol.ProtocolTest[jvm] > eventEmptyArgs[jvm] FAILED
    org.opentest4j.AssertionFailedError: expected:<"{"id":1,"tag":2[]}"> but was:<"{"id":1,"tag":2[,"args":[]]}">

app.cash.redwood.protocol.SnapshotChangeListTest[jvm] > validChangesSerialization[jvm] FAILED
    org.opentest4j.AssertionFailedError: expected:<...e":"modifier","id":1[]},{"type":"property"...> but was:<...e":"modifier","id":1[,"elements":[]]},{"type":"property"...>
```

The failure messages are janked up because assertk is mixing up its own display format with the JSON format, but the actual difference is the presence of e.g. `,"args":[]`.

**Comment by JakeWharton** [comment]
I suspect it's because it checks `.equals` against the default value which works for lists and fails for arrays.

I can write a custom serializer for those properties that omits it when empty.

**Comment by drewhamilton** [comment]
I pushed a serializer for `Event`, but it's very verbose. I couldn't find a way to supply a custom serializer for a single property, but I'm hoping there's a way.

If not, another option could be a constant `(internal?) val EmptyJsonElementArray` and stop using `emptyArray` and `arrayOf()` in generated code and tests.

**Comment by drewhamilton** [comment]
Trying out `SerializersModuleBuilder.contextual` but it's not getting picked up the way I hoped it would.

**Comment by JakeWharton** [comment]
Weird that it didn't retarget. I guess because it's from a fork?

**Comment by JakeWharton** [comment]
Back at a computer. It won't even let me re-open. Want to rebase and make a new PR when you get a chance?

**Pull request closed without merging**